core-tests/src/test/java/org/mule/runtime/core/internal/lock/InstanceLockGroupTestCase.java

```diff
@@ -57,6 +66,65 @@ public void testWhenSeveralLockOneUnlockThenDestroy() throws Exception {
     lockUnlockThenDestroy(5);
   }
 
+  @Test
+  @Issue("W-11929632")
+  public void disposeDoesNotLoseReferenceToTakenLocks() throws InterruptedException {
+    String testLockId = "TestLockId";
+    instanceLockGroup.lock(testLockId);
+    verify(lockProvider.getSpiedLock(testLockId)).lock();
+
+    // Spawn a task that disposes the lock group and wait some time to ensure that it actually started.
+    executor.submit(instanceLockGroup::dispose);
+    sleep(100);
```

Comment:
Honestly, the potential for flakiness from this worries me a little bit, but I'm not sure how we could avoid it without adding stuff to the `InstanceLockGroup` that would otherwise be unnecessary for productive code.
I'm worried because the delay needs to fall exactly in the sweet spot between two events: the call to `wait` inside the `dispose` method (because otherwise you wouldn't be testing anything) and the timeout expiration (because otherwise the assertion would fail).
What if we grabbed the `Thread` reference from the `Runnable` and then, on the main thread, add a probe which polls for the `Thread`'s state until it is waiting.
Then if it never gets to a waiting state after a reasonable period of time, which can be considerably bigger than the default timeout, we fail. The rationale would be that, if there are pending locks, the thread trying to run through `dispose` should wait (until the timeout at least).
It wouldn't be entirely bullet-proof but should be more robust, what do you think?

Reply:
Good idea, done